File: 298__Binary_Tree_Longest_Consecutive_Sequence.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def longestConsecutive(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.maxlength = 0
        def helper(node, parent, length):
            if node is None:
                return
            if parent is not None and node.val == parent.val + 1:
                length += 1
            else:
                length = 1
            self.maxlength = max(self.maxlength, length)
            helper(node.left, node, length)
            helper(node.right, node, length)
        helper(root, None, 0)
        return self.maxlength

File: Java/298__Binary_Tree_Longest_Consecutive_Sequence.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int longestConsecutive(TreeNode root) {
        return helper(root, null, 0);
    }

    private int helper(TreeNode node, TreeNode parent, int length) {
        if (node == null) return length;
        length = (parent != null && parent.val + 1 == node.val) ? length + 1 : 1;
        return Math.max(length, Math.max(helper(node.left, node, length), helper(node.right, node, length)));
    }
}
